src/rdkafka_header.c

```diff
@@ -0,0 +1,220 @@
+        hdrs->rkhdrs_ser_size = 0;
+
+        return hdrs;
+}
+
+static void *rd_kafka_header_copy (const void *_src, void *opaque) {
+        rd_kafka_headers_t *hdrs = opaque;
+        const rd_kafka_header_t *src = (const rd_kafka_header_t *)_src;
+
+        return (void *)rd_kafka_header_add(
+                hdrs,
+                src->rkhdr_name, src->rkhdr_name_size,
+                src->rkhdr_value, src->rkhdr_value_size);
+}
+
+rd_kafka_headers_t *
+rd_kafka_headers_copy (const rd_kafka_headers_t *src) {
+        rd_kafka_headers_t *dst;
+
+        dst = rd_malloc(sizeof(*dst));
+        rd_list_init(&dst->rkhdrs_list, rd_list_cnt(&src->rkhdrs_list),
+                     rd_kafka_header_destroy);
+        dst->rkhdrs_ser_size = 0; /* Updated by header_copy() */
+        rd_list_copy_to(&dst->rkhdrs_list, &src->rkhdrs_list,
+                        rd_kafka_header_copy, dst);
+
+        return dst;
+}
+
+
+
+rd_kafka_resp_err_t
+rd_kafka_header_add (rd_kafka_headers_t *hdrs,
+                     const char *name, ssize_t name_size,
+                     const void *value, ssize_t value_size) {
+        rd_kafka_header_t *hdr;
+        char varint_NameLen[RD_UVARINT_ENC_SIZEOF(int32_t)];
+        char varint_ValueLen[RD_UVARINT_ENC_SIZEOF(int32_t)];
+
+        if (name_size == -1)
+                name_size = strlen(name);
+
+        if (value_size == -1)
+                value_size = value ? strlen(value) : 0;
+        else if (!value)
+                value_size = 0;
+
+        hdr = rd_malloc(sizeof(*hdr) + name_size + 1 + value_size + 1);
+        hdr->rkhdr_name_size = name_size;
+        memcpy((void *)hdr->rkhdr_name, name, name_size);
+        hdr->rkhdr_name[name_size] = '\0';
+
+        if (likely(value != NULL)) {
+                hdr->rkhdr_value = hdr->rkhdr_name+name_size+1;
+                memcpy((void *)hdr->rkhdr_value, value, value_size);
+                hdr->rkhdr_value[value_size] = '\0';
+                hdr->rkhdr_value_size        = value_size;
+        } else {
+                hdr->rkhdr_value      = NULL;
+                hdr->rkhdr_value_size = 0;
+        }
+
+        rd_list_add(&hdrs->rkhdrs_list, hdr);
+
+        /* Calculate serialized size of header */
+        hdr->rkhdr_ser_size = name_size + value_size;
+        hdr->rkhdr_ser_size += rd_uvarint_enc_i64(varint_NameLen,
+                                                  sizeof(varint_NameLen),
+                                                  name_size);
+        hdr->rkhdr_ser_size += rd_uvarint_enc_i64(varint_ValueLen,
+                                                  sizeof(varint_ValueLen),
+                                                  value_size);
+        hdrs->rkhdrs_ser_size += hdr->rkhdr_ser_size;
+
+        return RD_KAFKA_RESP_ERR_NO_ERROR;
+}
+
+
+/**
+ * @brief header_t(name) to char * comparator
+ */
+static int rd_kafka_header_cmp_str (void *_a, void *_b) {
+        const rd_kafka_header_t *a = _a;
```

Comment:
not sure why you have `void*` parameters then immediately cast?

Reply:
The comparator signature is `int () (void *, void *)`, passing a function with another signature would cause a compiler warning